File: java/advanced/Graph.java
package advanced;

import java.util.*;

/**
 * Directed or undirected weighted graph using an adjacency list representation.
 *
 * <p>Vertices are identified by generic labels (type {@code V}).
 * Edges carry {@code double} weights; use weight 1.0 for unweighted graphs.
 *
 * <p>Supported operations and complexities (V = vertices, E = edges):
 * <ul>
 *   <li>addVertex       – O(1) amortized</li>
 *   <li>addEdge         – O(1) amortized</li>
 *   <li>removeEdge      – O(degree(u))</li>
 *   <li>bfs             – O(V + E)</li>
 *   <li>dfs             – O(V + E)</li>
 *   <li>hasCycle        – O(V + E)</li>
 *   <li>topologicalSort – O(V + E) (Kahn's algorithm, directed only)</li>
 *   <li>shortestPath    – O((V + E) log V) Dijkstra with binary heap</li>
 * </ul>
 *
 * <p>Space complexity: O(V + E).
 *
 * @param <V> vertex label type (must implement {@link Comparable} for deterministic output)
 */
public class Graph<V extends Comparable<V>> {

    // -------------------------------------------------------------------------
    // Edge helper
    // -------------------------------------------------------------------------

    /** Represents a weighted directed edge to {@code dest}. */
    private static class Edge<V> {
        V dest;
        double weight;

        Edge(V dest, double weight) {
            this.dest = dest;
            this.weight = weight;
        }

        @Override
        public String toString() {
            return dest + "(" + weight + ")";
        }
    }

    // -------------------------------------------------------------------------
    // Fields
    // -------------------------------------------------------------------------

    /** Adjacency list: vertex -> list of outgoing edges. */
    private final Map<V, List<Edge<V>>> adjList;
    private final boolean directed;
    private int edgeCount;

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    /**
     * Creates a new graph.
     *
     * @param directed {@code true} for a directed graph, {@code false} for undirected
     */
    public Graph(boolean directed) {
        this.directed = directed;
        adjList = new LinkedHashMap<>();
    }

    // -------------------------------------------------------------------------
    // addVertex
    // -------------------------------------------------------------------------

    /**
     * Adds a vertex to the graph.  If the vertex already exists this is a no-op.
     *
     * <p>Time: O(1) amortized.
     *
     * @param v the vertex label
     */
    public void addVertex(V v) {
        adjList.putIfAbsent(v, new ArrayList<>());
    }

    // -------------------------------------------------------------------------
    // addEdge
    // -------------------------------------------------------------------------

    /**
     * Adds a weighted edge between {@code u} and {@code v}.
     * For undirected graphs a reverse edge is also added.
     * Both vertices are created if they do not already exist.
     *
     * <p>Time: O(1) amortized.
     *
     * @param u      source vertex
     * @param v      destination vertex
     * @param weight edge weight
     */
    public void addEdge(V u, V v, double weight) {
        addVertex(u);
        addVertex(v);
        adjList.get(u).add(new Edge<>(v, weight));
        edgeCount++;
        if (!directed) {
            adjList.get(v).add(new Edge<>(u, weight));
        }
    }

    /**
     * Convenience overload: adds an unweighted edge (weight = 1.0).
     *
     * @param u source vertex
     * @param v destination vertex
     */
    public void addEdge(V u, V v) {
        addEdge(u, v, 1.0);
    }

    // -------------------------------------------------------------------------
    // removeEdge
    // -------------------------------------------------------------------------

    /**
     * Removes the edge from {@code u} to {@code v} (and the reverse for undirected).
     *
     * <p>Time: O(degree(u)) [ + O(degree(v)) for undirected ].
     *
     * @param u source
     * @param v destination
     * @return {@code true} if the edge existed and was removed
     */
    public boolean removeEdge(V u, V v) {
        List<Edge<V>> uList = adjList.get(u);
        if (uList == null) return false;
        boolean removed = uList.removeIf(e -> e.dest.equals(v));
        if (removed) {
            edgeCount--;
            if (!directed) {
                adjList.get(v).removeIf(e -> e.dest.equals(u));
            }
        }
        return removed;
    }

    // -------------------------------------------------------------------------
    // BFS
    // -------------------------------------------------------------------------

    /**
     * Returns the vertices reachable from {@code start} in BFS order.
     *
     * <p>Time: O(V + E) | Space: O(V).
     *
     * @param start the starting vertex
     * @return BFS traversal order
     */
    public List<V> bfs(V start) {
        List<V> order = new ArrayList<>();
        if (!adjList.containsKey(start)) return order;

        Set<V> visited = new LinkedHashSet<>();
        Queue<V> queue = new LinkedList<>();

        visited.add(start);
        queue.offer(start);

        while (!queue.isEmpty()) {
            V cur = queue.poll();
            order.add(cur);
            List<Edge<V>> neighbors = adjList.getOrDefault(cur, Collections.emptyList());
            // Sort neighbors for deterministic output
            neighbors.stream()
                     .map(e -> e.dest)
                     .filter(n -> !visited.contains(n))
                     .sorted()
                     .forEach(n -> { visited.add(n); queue.offer(n); });
        }
        return order;
    }

    // -------------------------------------------------------------------------
    // DFS
    // -------------------------------------------------------------------------

    /**
     * Returns the vertices reachable from {@code start} in DFS order.
     *
     * <p>Time: O(V + E) | Space: O(V).
     *
     * @param start the starting vertex
     * @return DFS traversal order
     */
    public List<V> dfs(V start) {
        List<V> order = new ArrayList<>();
        if (!adjList.containsKey(start)) return order;
        Set<V> visited = new LinkedHashSet<>();
        dfsRec(start, visited, order);
        return order;
    }

    private void dfsRec(V cur, Set<V> visited, List<V> order) {
        visited.add(cur);
        order.add(cur);
        adjList.getOrDefault(cur, Collections.emptyList())
               .stream()
               .map(e -> e.dest)
               .filter(n -> !visited.contains(n))
               .sorted()
               .forEach(n -> dfsRec(n, visited, order));
    }

    // -------------------------------------------------------------------------
    // hasCycle
    // -------------------------------------------------------------------------

    /**
     * Returns {@code true} if the graph contains at least one cycle.
     *
     * <p>For <em>directed</em> graphs uses DFS with a recursion stack (WHITE-GRAY-BLACK coloring).
     * For <em>undirected</em> graphs uses DFS tracking the parent to avoid trivial back-edges.
     *
     * <p>Time: O(V + E) | Space: O(V).
     *
     * @return {@code true} if a cycle exists
     */
    public boolean hasCycle() {
        Set<V> visited = new HashSet<>();
        if (directed) {
            Set<V> recStack = new HashSet<>();
            for (V v : adjList.keySet()) {
                if (!visited.contains(v) && hasCycleDirected(v, visited, recStack)) return true;
            }
        } else {
            for (V v : adjList.keySet()) {
                if (!visited.contains(v) && hasCycleUndirected(v, null, visited)) return true;
            }
        }
        return false;
    }

    private boolean hasCycleDirected(V cur, Set<V> visited, Set<V> recStack) {
        visited.add(cur);
        recStack.add(cur);
        for (Edge<V> e : adjList.getOrDefault(cur, Collections.emptyList())) {
            if (!visited.contains(e.dest)) {
                if (hasCycleDirected(e.dest, visited, recStack)) return true;
            } else if (recStack.contains(e.dest)) {
                return true;
            }
        }
        recStack.remove(cur);
        return false;
    }

    private boolean hasCycleUndirected(V cur, V parent, Set<V> visited) {
        visited.add(cur);
        for (Edge<V> e : adjList.getOrDefault(cur, Collections.emptyList())) {
            if (!visited.contains(e.dest)) {
                if (hasCycleUndirected(e.dest, cur, visited)) return true;
            } else if (!e.dest.equals(parent)) {
                return true;
            }
        }
        return false;
    }

    // -------------------------------------------------------------------------
    // Topological Sort (Kahn's algorithm – directed graphs only)
    // -------------------------------------------------------------------------

    /**
     * Returns a topological ordering of the vertices (directed acyclic graphs only).
     *
     * <p>Uses Kahn's BFS-based algorithm.  If the graph has a cycle, returns an
     * empty list (not all vertices can be ordered).
     *
     * <p>Time: O(V + E) | Space: O(V).
     *
     * @return topological order, or empty list if a cycle exists / graph is undirected
     */
    public List<V> topologicalSort() {
        if (!directed) throw new UnsupportedOperationException("Topological sort requires a directed graph");

        // Compute in-degrees
        Map<V, Integer> inDegree = new LinkedHashMap<>();
        for (V v : adjList.keySet()) inDegree.put(v, 0);
        for (V u : adjList.keySet()) {
            for (Edge<V> e : adjList.get(u)) {
                inDegree.merge(e.dest, 1, Integer::sum);
            }
        }

        // Enqueue all zero-in-degree vertices (sorted for determinism)
        PriorityQueue<V> pq = new PriorityQueue<>();
        for (Map.Entry<V, Integer> entry : inDegree.entrySet()) {
            if (entry.getValue() == 0) pq.offer(entry.getKey());
        }

        List<V> order = new ArrayList<>();
        while (!pq.isEmpty()) {
            V cur = pq.poll();
            order.add(cur);
            for (Edge<V> e : adjList.getOrDefault(cur, Collections.emptyList())) {
                int newDeg = inDegree.merge(e.dest, -1, Integer::sum);
                if (newDeg == 0) pq.offer(e.dest);
            }
        }

        return order.size() == adjList.size() ? order : Collections.emptyList();
    }

    // -------------------------------------------------------------------------
    // Dijkstra's Shortest Path
    // -------------------------------------------------------------------------

    /**
     * Computes the shortest (minimum weight) path from {@code source} to all
     * other reachable vertices using Dijkstra's algorithm.
     *
     * <p>Assumes non-negative edge weights.
     *
     * <p>Time: O((V + E) log V) | Space: O(V).
     *
     * @param source the starting vertex
     * @return map from vertex to its shortest distance from {@code source};
     *         unreachable vertices map to {@link Double#POSITIVE_INFINITY}
     */
    public Map<V, Double> shortestPath(V source) {
        Map<V, Double> dist  = new HashMap<>();
        Map<V, V>      prev  = new HashMap<>();
        // Min-heap: [distance, vertex] – using a simple wrapper pair
        PriorityQueue<double[]> pq = new PriorityQueue<>(Comparator.comparingDouble(a -> a[0]));
        // We encode vertex index via a parallel index map
        List<V> vertices = new ArrayList<>(adjList.keySet());
        Map<V, Integer> indexMap = new HashMap<>();
        for (int i = 0; i < vertices.size(); i++) indexMap.put(vertices.get(i), i);

        for (V v : adjList.keySet()) dist.put(v, Double.POSITIVE_INFINITY);
        dist.put(source, 0.0);
        pq.offer(new double[]{0.0, indexMap.getOrDefault(source, -1)});

        while (!pq.isEmpty()) {
            double[] top = pq.poll();
            double d = top[0];
            int idx = (int) top[1];
            if (idx < 0 || idx >= vertices.size()) continue;
            V cur = vertices.get(idx);
            if (d > dist.get(cur)) continue; // stale entry

            for (Edge<V> e : adjList.getOrDefault(cur, Collections.emptyList())) {
                double newDist = dist.get(cur) + e.weight;
                if (newDist < dist.getOrDefault(e.dest, Double.POSITIVE_INFINITY)) {
                    dist.put(e.dest, newDist);
                    prev.put(e.dest, cur);
                    pq.offer(new double[]{newDist, indexMap.getOrDefault(e.dest, -1)});
                }
            }
        }
        return dist;
    }

    /**
     * Returns the shortest path (list of vertices) from {@code source} to {@code dest}.
     *
     * <p>Time: O((V + E) log V) | Space: O(V).
     *
     * @param source start
     * @param dest   end
     * @return ordered list of vertices on the shortest path, or empty list if unreachable
     */
    public List<V> shortestPathTo(V source, V dest) {
        // Re-run Dijkstra tracking predecessors
        Map<V, Double> dist = new HashMap<>();
        Map<V, V>      prev = new HashMap<>();
        List<V> vertices = new ArrayList<>(adjList.keySet());
        Map<V, Integer> indexMap = new HashMap<>();
        for (int i = 0; i < vertices.size(); i++) indexMap.put(vertices.get(i), i);

        for (V v : adjList.keySet()) dist.put(v, Double.POSITIVE_INFINITY);
        dist.put(source, 0.0);
        PriorityQueue<double[]> pq = new PriorityQueue<>(Comparator.comparingDouble(a -> a[0]));
        pq.offer(new double[]{0.0, indexMap.getOrDefault(source, -1)});

        while (!pq.isEmpty()) {
            double[] top = pq.poll();
            double d = top[0];
            int idx = (int) top[1];
            if (idx < 0 || idx >= vertices.size()) continue;
            V cur = vertices.get(idx);
            if (d > dist.get(cur)) continue;
            for (Edge<V> e : adjList.getOrDefault(cur, Collections.emptyList())) {
                double nd = dist.get(cur) + e.weight;
                if (nd < dist.getOrDefault(e.dest, Double.POSITIVE_INFINITY)) {
                    dist.put(e.dest, nd);
                    prev.put(e.dest, cur);
                    pq.offer(new double[]{nd, indexMap.getOrDefault(e.dest, -1)});
                }
            }
        }

        // Reconstruct path
        LinkedList<V> path = new LinkedList<>();
        V cur = dest;
        while (cur != null) {
            path.addFirst(cur);
            cur = prev.get(cur);
        }
        return path.getFirst().equals(source) ? path : Collections.emptyList();
    }

    // -------------------------------------------------------------------------
    // Accessors
    // -------------------------------------------------------------------------

    /** Returns the number of vertices. */
    public int vertexCount() { return adjList.size(); }

    /** Returns the number of edges (directed count for directed graphs). */
    public int edgeCount() { return edgeCount; }

    /** Returns true if the graph is directed. */
    public boolean isDirected() { return directed; }

    // -------------------------------------------------------------------------
    // toString – adjacency list view
    // -------------------------------------------------------------------------

    /**
     * Returns a human-readable adjacency list representation.
     *
     * <p>Time: O(V + E) | Space: O(V + E).
     *
     * @return adjacency list string
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(directed ? "Directed" : "Undirected")
          .append(" Graph (V=").append(vertexCount())
          .append(", E=").append(edgeCount).append("):\n");
        for (Map.Entry<V, List<Edge<V>>> entry : adjList.entrySet()) {
            sb.append("  ").append(entry.getKey()).append(" -> ")
              .append(entry.getValue()).append("\n");
        }
        return sb.toString();
    }

    // -------------------------------------------------------------------------
    // Main – demo
    // -------------------------------------------------------------------------

    public static void main(String[] args) {
        System.out.println("=== Graph Demo ===\n");

        // --- Directed weighted graph ---
        Graph<String> dg = new Graph<>(true);
        dg.addEdge("A", "B", 4);
        dg.addEdge("A", "C", 2);
        dg.addEdge("B", "C", 5);
        dg.addEdge("B", "D", 10);
        dg.addEdge("C", "E", 3);
        dg.addEdge("E", "D", 4);
        dg.addEdge("D", "F", 11);

        System.out.println(dg);
        System.out.println("BFS from A  : " + dg.bfs("A"));
        System.out.println("DFS from A  : " + dg.dfs("A"));
        System.out.println("Has cycle   : " + dg.hasCycle());
        System.out.println("Topo sort   : " + dg.topologicalSort());

        Map<String, Double> dist = dg.shortestPath("A");
        System.out.println("\nDijkstra from A:");
        new TreeMap<>(dist).forEach((v, d) -> System.out.println("  A -> " + v + " : " + d));

        System.out.println("\nShortest path A -> F : " + dg.shortestPathTo("A", "F"));

        // --- Undirected graph with cycle ---
        Graph<Integer> ug = new Graph<>(false);
        ug.addEdge(1, 2);
        ug.addEdge(2, 3);
        ug.addEdge(3, 4);
        ug.addEdge(4, 2); // creates a cycle 2-3-4-2
        System.out.println("\n" + ug);
        System.out.println("Has cycle   : " + ug.hasCycle());
        System.out.println("BFS from 1  : " + ug.bfs(1));
        System.out.println("DFS from 1  : " + ug.dfs(1));

        // removeEdge demo
        System.out.println("\nRemove edge 4-2...");
        ug.removeEdge(4, 2);
        System.out.println("Has cycle after removal: " + ug.hasCycle());
        System.out.println(ug);

        // DAG topological sort
        Graph<String> dag = new Graph<>(true);
        dag.addEdge("wash", "dry");
        dag.addEdge("wash", "fold");
        dag.addEdge("dry",  "fold");
        dag.addEdge("fold", "put away");
        System.out.println("\nDAG topo sort (laundry): " + dag.topologicalSort());
    }
}

File: python/advanced/graph.py
"""
Graph (Directed & Undirected) — Adjacency List Representation
==============================================================
Supports both directed and undirected graphs with unweighted edges.
Vertices can be any hashable type.

Complexities  (V = vertices, E = edges):
    - add_vertex / add_edge / remove_edge: O(1) amortized
    - BFS / DFS:                           O(V + E)
    - has_cycle (undirected):              O(V + E)
    - has_cycle (directed, DFS color):     O(V + E)
    - topological_sort:                    O(V + E)
    - shortest_path (BFS, unweighted):     O(V + E)
    - Space:                               O(V + E)
"""

from __future__ import annotations
from collections import deque, defaultdict
from typing import Any, Dict, List, Optional, Set


class Graph:
    """
    Adjacency-list graph.

    Parameters
    ----------
    directed : bool
        If True, edges are one-directional (u → v).
        If False (default), each edge is stored in both directions.
    """

    def __init__(self, directed: bool = False) -> None:
        self._directed = directed
        # adjacency list: vertex → set of neighbours
        self._adj: Dict[Any, Set[Any]] = defaultdict(set)

    # ------------------------------------------------------------------
    # Mutation
    # ------------------------------------------------------------------

    def add_vertex(self, v: Any) -> None:
        """
        Ensure *v* exists in the graph (creates an isolated vertex if new).

        Time:  O(1)
        """
        if v not in self._adj:
            self._adj[v] = set()

    def add_edge(self, u: Any, v: Any) -> None:
        """
        Add an edge u → v (and v → u for undirected graphs).
        Auto-creates vertices if they don't exist.

        Time:  O(1)
        """
        self._adj[u].add(v)
        if not self._directed:
            self._adj[v].add(u)
        else:
            # Make sure v exists as a vertex even if it has no outgoing edges.
            if v not in self._adj:
                self._adj[v] = set()

    def remove_edge(self, u: Any, v: Any) -> None:
        """
        Remove the edge u → v (no-op if it doesn't exist).

        Time:  O(1)
        """
        self._adj[u].discard(v)
        if not self._directed:
            self._adj[v].discard(u)

    @property
    def vertices(self) -> List[Any]:
        return list(self._adj.keys())

    @property
    def edges(self) -> List[tuple]:
        """Return all edges as (u, v) tuples (each undirected edge once)."""
        seen: Set[frozenset] = set()
        result = []
        for u, neighbours in self._adj.items():
            for v in neighbours:
                key = frozenset([u, v]) if not self._directed else (u, v)
                if key not in seen:
                    seen.add(key)
                    result.append((u, v))
        return result

    # ------------------------------------------------------------------
    # BFS
    # ------------------------------------------------------------------

    def bfs(self, start: Any) -> List[Any]:
        """
        Breadth-first traversal from *start*.
        Returns vertices in the order they are first visited.

        Time:  O(V + E)
        Space: O(V)
        """
        if start not in self._adj:
            return []
        visited: Set[Any] = {start}
        order: List[Any] = []
        queue: deque = deque([start])
        while queue:
            node = queue.popleft()
            order.append(node)
            for neighbour in sorted(self._adj[node]):  # sorted for determinism
                if neighbour not in visited:
                    visited.add(neighbour)
                    queue.append(neighbour)
        return order

    # ------------------------------------------------------------------
    # DFS
    # ------------------------------------------------------------------

    def dfs(self, start: Any) -> List[Any]:
        """
        Depth-first traversal from *start* (iterative).
        Returns vertices in the order they are first visited.

        Time:  O(V + E)
        Space: O(V)
        """
        if start not in self._adj:
            return []
        visited: Set[Any] = set()
        order: List[Any] = []
        stack = [start]
        while stack:
            node = stack.pop()
            if node in visited:
                continue
            visited.add(node)
            order.append(node)
            for neighbour in sorted(self._adj[node], reverse=True):
                if neighbour not in visited:
                    stack.append(neighbour)
        return order

    # ------------------------------------------------------------------
    # Cycle detection
    # ------------------------------------------------------------------

    def has_cycle(self) -> bool:
        """
        Detect whether the graph contains a cycle.

        Undirected: DFS with parent tracking — back edge ⟹ cycle.
        Directed:   DFS three-colour algorithm (white/grey/black).
                    grey node reached again ⟹ cycle.

        Time:  O(V + E)
        Space: O(V)
        """
        if self._directed:
            return self._has_cycle_directed()
        return self._has_cycle_undirected()

    def _has_cycle_undirected(self) -> bool:
        visited: Set[Any] = set()

        def dfs(node: Any, parent: Any) -> bool:
            visited.add(node)
            for nb in self._adj[node]:
                if nb not in visited:
                    if dfs(nb, node):
                        return True
                elif nb != parent:
                    return True  # back edge to a non-parent ancestor
            return False

        for v in self._adj:
            if v not in visited:
                if dfs(v, None):
                    return True
        return False

    def _has_cycle_directed(self) -> bool:
        # 0 = white (unvisited), 1 = grey (in stack), 2 = black (done)
        color: Dict[Any, int] = {v: 0 for v in self._adj}

        def dfs(node: Any) -> bool:
            color[node] = 1  # grey
            for nb in self._adj[node]:
                if color[nb] == 1:
                    return True  # back edge
                if color[nb] == 0 and dfs(nb):
                    return True
            color[node] = 2  # black
            return False

        return any(dfs(v) for v in self._adj if color[v] == 0)

    # ------------------------------------------------------------------
    # Topological sort (Kahn's BFS-based algorithm)
    # ------------------------------------------------------------------

    def topological_sort(self) -> List[Any]:
        """
        Return a topological ordering of vertices (directed graphs only).
        Returns an empty list if the graph is not a DAG (cycle detected).

        Uses Kahn's algorithm (iterative, BFS-based).

        Time:  O(V + E)
        Space: O(V)
        """
        if not self._directed:
            raise ValueError("Topological sort is only defined for directed graphs.")

        in_degree: Dict[Any, int] = {v: 0 for v in self._adj}
        for v in self._adj:
            for nb in self._adj[v]:
                in_degree[nb] += 1

        queue: deque = deque(sorted(v for v, d in in_degree.items() if d == 0))
        order: List[Any] = []

        while queue:
            node = queue.popleft()
            order.append(node)
            for nb in sorted(self._adj[node]):
                in_degree[nb] -= 1
                if in_degree[nb] == 0:
                    queue.append(nb)

        if len(order) != len(self._adj):
            return []  # cycle detected
        return order

    # ------------------------------------------------------------------
    # Shortest path (BFS, unweighted)
    # ------------------------------------------------------------------

    def shortest_path(self, start: Any, end: Any) -> Optional[List[Any]]:
        """
        Return the shortest path from *start* to *end* as a list of vertices,
        or None if no path exists.

        Works for both directed and undirected unweighted graphs.
        For weighted graphs, use Dijkstra's algorithm instead.

        Time:  O(V + E)
        Space: O(V)
        """
        if start not in self._adj or end not in self._adj:
            return None
        if start == end:
            return [start]

        visited: Set[Any] = {start}
        # Store full path in queue to simplify reconstruction.
        queue: deque = deque([[start]])

        while queue:
            path = queue.popleft()
            node = path[-1]
            for nb in sorted(self._adj[node]):
                if nb == end:
                    return path + [nb]
                if nb not in visited:
                    visited.add(nb)
                    queue.append(path + [nb])
        return None

    # ------------------------------------------------------------------
    # String representation
    # ------------------------------------------------------------------

    def __str__(self) -> str:
        kind = "Directed" if self._directed else "Undirected"
        lines = [f"{kind} Graph (V={len(self._adj)}, E={len(self.edges)}):"]
        for v in sorted(self._adj.keys(), key=str):
            neighbours = ", ".join(str(nb) for nb in sorted(self._adj[v], key=str))
            arrow = "→" if self._directed else "—"
            lines.append(f"  {v} {arrow} [{neighbours}]")
        return "\n".join(lines)


# ----------------------------------------------------------------------
# Demo
# ----------------------------------------------------------------------

if __name__ == "__main__":
    print("=== Undirected Graph ===")
    g = Graph(directed=False)
    edges = [(1, 2), (1, 3), (2, 4), (3, 4), (4, 5)]
    for u, v in edges:
        g.add_edge(u, v)
    print(g)
    print("BFS from 1:", g.bfs(1))
    print("DFS from 1:", g.dfs(1))
    print("Has cycle? :", g.has_cycle())
    print("Shortest path 1→5:", g.shortest_path(1, 5))

    print()
    print("=== Directed Graph (DAG) ===")
    dag = Graph(directed=True)
    dag_edges = [
        ("A", "B"), ("A", "C"), ("B", "D"),
        ("C", "D"), ("D", "E"), ("B", "E"),
    ]
    for u, v in dag_edges:
        dag.add_edge(u, v)
    print(dag)
    print("BFS from A    :", dag.bfs("A"))
    print("DFS from A    :", dag.dfs("A"))
    print("Has cycle?    :", dag.has_cycle())
    print("Topo sort     :", dag.topological_sort())
    print("Shortest A→E  :", dag.shortest_path("A", "E"))

    print()
    print("=== Directed Graph with Cycle ===")
    cyc = Graph(directed=True)
    cyc.add_edge(1, 2)
    cyc.add_edge(2, 3)
    cyc.add_edge(3, 1)  # cycle
    print(cyc)
    print("Has cycle?    :", cyc.has_cycle())
    print("Topo sort     :", cyc.topological_sort(), "(empty = cycle detected)")
